File: backend/src/main/java/com/venus/back_end/service/FCDService.java
package com.venus.back_end.service;

import com.venus.back_end.entity.Product;
import com.venus.back_end.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class FCDService {

    @Value("${fcd.api.key}")
    private String apiKey;

    @Value("${fcd.api.url}")
    private String apiUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    @Autowired
    private ProductRepository productRepository;

    private String getUrl(String command) {
        return UriComponentsBuilder.fromHttpUrl(apiUrl + command)
                .queryParam("api_key", apiKey)
                .toUriString();
    }

    public List<Map<String, Object>> find(String name) {
        String url = getUrl("foods/search") + "&query=" + name;
        Map<String, Object> response = restTemplate.getForObject(url, Map.class);
        return (List<Map<String, Object>>) response.get("foods");
    }

    public Map<String, Object> getReport(String ndbno) {
        String url = getUrl("food/" + ndbno);
        return restTemplate.getForObject(url, Map.class);
    }

    public List<Map<String, Object>> getNutrients(String ndbno) {
        Map<String, Object> report = getReport(ndbno);
        return (List<Map<String, Object>>) report.get("foodNutrients");
    }

    public Map<String, Double> getMeasures(String ndbno) {
        Map<String, Object> report = getReport(ndbno);
        List<Map<String, Object>> portions = (List<Map<String, Object>>) report.get("foodPortions");
        Map<String, Double> measures = new HashMap<>();
        for (Map<String, Object> portion : portions) {
            String portionDescription = (String) portion.get("portionDescription");
            if (portionDescription != null) {
                measures.put(portionDescription, ((Number) portion.get("gramWeight")).doubleValue());
            }
        }
        return measures;
    }

    public List<Map<String, Object>> calculateNutrients(String ndbno, String portionDescription, double quantity) {
        Map<String, Object> report = getReport(ndbno);
        List<Map<String, Object>> nutrients = (List<Map<String, Object>>) report.get("foodNutrients");
        List<Map<String, Object>> portions = (List<Map<String, Object>>) report.get("foodPortions");

        Map<String, Object> chosenPortion = portions.stream()
                .filter(portion -> portionDescription.equals(portion.get("portionDescription")))
                .findFirst()
                .orElse(null);

        if (chosenPortion == null) {
            throw new IllegalArgumentException("Portion description not found");
        }

        double chosenGramWeight = ((Number) chosenPortion.get("gramWeight")).doubleValue();

        for (Map<String, Object> nutrient : nutrients) {
            double amountPer100g = ((Number) nutrient.get("amount")).doubleValue();
            double amountPerGram = amountPer100g / 100;
            double amountForChosenPortion = amountPerGram * chosenGramWeight * quantity;
            nutrient.put("intake", amountForChosenPortion);
        }

        return nutrients;
    }

    public List<Map<String, Object>> calculateNutrientsGram(String ndbno, double grams) {
        Map<String, Object> report = getReport(ndbno);
        List<Map<String, Object>> nutrients = (List<Map<String, Object>>) report.get("foodNutrients");

        for (Map<String, Object> nutrient : nutrients) {
            double amountPer100g = ((Number) nutrient.get("amount")).doubleValue();
            double amountPerGram = amountPer100g / 100;
            double amountForGivenGrams = amountPerGram * grams;
            nutrient.put("intake", amountForGivenGrams);
        }

        return nutrients;
    }

    public Map<String, Double> calculateDailyNutrients(LocalDate date, List<Product> products) {
        Map<String, Double> totalNutrients = new HashMap<>();
        if (products == null || products.isEmpty()) {
            return totalNutrients;
        }


        for (Product product : products) {
            String ndbno = product.getFdcId();
            double quantity = product.getQuantity();
            double gram = product.getGram();
            String portionDescription = product.getPortion();

            List<Map<String, Object>> nutrients;
            if (gram > 0) {
                nutrients = calculateNutrientsGram(ndbno, gram * quantity);
            } else if (portionDescription != null) {
                nutrients = calculateNutrients(ndbno, portionDescription, quantity);
            } else {
                throw new IllegalArgumentException("Invalid product data");
            }

            for (Map<String, Object> nutrient : nutrients) {
                Map<String, Object> nutrientDetails = (Map<String, Object>) nutrient.get("nutrient");
                String label = (String) nutrientDetails.get("name");
                if (label != null) {
                    double intake = ((Number) nutrient.get("intake")).doubleValue();
                    totalNutrients.put(label, totalNutrients.getOrDefault(label, 0.0) + intake);
                }
            }
        }
        return totalNutrients;
    }
}

File: frontend/src/services/FCDService.js
import axios from "axios";
import AuthService from "./AuthService"

class FCDService {
  static API_URL = "http://localhost:8080";
  static cache = {};

  static async find(name) {
    const url = `${this.API_URL}/find?name=${name}`;
    const response = await axios.get(url, {headers: {
      'Authorization': `Bearer ${AuthService.getToken()}`
    }});
    return response.data;
  }

  static async get_nutrients(ndbno) {
    const url = `${this.API_URL}/nutrients/${String(ndbno)}`;
    const response = await axios.get(url, {headers: {
      'Authorization': `Bearer ${AuthService.getToken()}`
    }});
    return response.data;
  }

  static async get_measures(ndbno) {
    const url = `${this.API_URL}/measures/${String(ndbno)}`;
    const response = await axios.get(url, {headers: {
      'Authorization': `Bearer ${AuthService.getToken()}`
    }});
    return response.data;
  }

  static async calculate_nutrients(ndbno, portion_description, quantity) {
    const url = `${this.API_URL}/calculate-nutrients`;
    const response = await axios.post(url, {
      ndbno: String(ndbno),
      portion_description,
      quantity,
      headers: {
        'Authorization': `Bearer ${AuthService.getToken()}`
      }
    });
    return response.data;
  }

  static async calculate_nutrients_gram(ndbno, grams) {
    const url = `${this.API_URL}/calculate-nutrients-gram`;
    const response = await axios.post(url, {
      ndbno: String(ndbno),
      grams,
      headers: {
        'Authorization': `Bearer ${AuthService.getToken()}`
      }
    });
    return response.data;
  }

  static async fetchNutrientsForProducts(products, selectedProduct) {
    const nutrientPromises = products.map(async (product) => {
      let nutrientsArr;
      if (product.gram > 0) {
        nutrientsArr = await this.calculate_nutrients_gram(
            String(product.fdcId),
            product.gram * product.quantity
        );
      } else if (product.portion) {
        nutrientsArr = await this.calculate_nutrients(
            String(product.fdcId),
            product.portion,
            product.quantity
        );
      } else {
        const servingSize = product.gram || selectedProduct.servingSize;
        nutrientsArr = await this.calculate_nutrients_gram(
            String(product.fdcId),
            servingSize
        );
      }
      return nutrientsArr;
    });

    const allNutrients = await Promise.all(nutrientPromises);
    const totalNutrients = allNutrients.flat().reduce((acc, curr) => {
      if (acc[curr.nutrient.name]) {
        acc[curr.nutrient.name] += curr.intake;
      } else {
        acc[curr.nutrient.name] = curr.intake;
      }
      return acc;
    }, {});
    return totalNutrients;
  }

  static async calculateDailyNutrients(date) {
    // Check cache
    if (this.cache[date]) {
      return this.cache[date];
    }

    const response = await axios.get(`${this.API_URL}/calculate-daily-nutrients`, {
      params: { date },
      headers: {
        'Authorization': `Bearer ${AuthService.getToken()}`
      }
    });

    // Store in cache
    this.cache[date] = response.data;
    return response.data;
  }

  static async invalidateCache(date) {
    if (this.cache[date]) {
      delete this.cache[date];
    }
  }
}

export default FCDService;
